Support for MFEMTransient in TransientMultiApps

alexanderianblair opened the pull request:
Closes #31017 

Would prefer to avoid multiple inheritance altogether and repurpose the old `MFEMExecutioner` class into something akin to a solve object... but this would need moving some additional logic to `ProblemOperators` which is likely outside the scope of this PR.

alexanderianblair committed:
Rename attributes of MFEMTransient to more closely match those in TransientBase

alexanderianblair committed:
Refactor MFEMTransient::step to support keepGoing termination check, for closer alignment with TransientBase

alexanderianblair committed:
Add MFEMTransient::incrementStepOrReject method

alexanderianblair committed:
Change inheritance tree for MFEMExecutioners

alexanderianblair committed:
Make MFEMTransient a child class of TransientBase

alexanderianblair committed:
Override fewer TransientBase methods in MFEMTransient

alexanderianblair committed:
Add tests for timestepper usage with MFEMTransient

alexanderianblair committed:
Renamce MFEM timestepper tests

alexanderianblair committed:
Add test for usage of TransientMultiApp with MFEMTransient executioner

alexanderianblair committed:
Add missing square.e input mesh

alexanderianblair committed:
Revert "Add test for usage of TransientMultiApp with MFEMTransient executioner" due to incorrect number of timesteps in sub app output

This reverts commit 42694cdffd45ff027ae5a0d90ce4eb66dae0d2cf.

alexanderianblair committed:
Move relevant steps from FixedPointSolve::solve() into MFEMTransient that were otherwise missing

alexanderianblair committed:
Call MFEMExecutioner::solve() in MFEMSteady

alexanderianblair committed:
Add tests for MultiApp usage with MFEM transient executioners

alexanderianblair committed:
Remove trailing whitespace from MFEM multiapp tests. Refs #31017

alexanderianblair committed:
Remove now redundant MFEMTransient output parameters.. Refs #31017

alexanderianblair committed:
Add capabilities to MFEM timesteppers tests. Refs #31017

alexanderianblair committed:
Fix documentation reference in mfem/multiapps/tests. Refs #31017

alexanderianblair committed:
Fix indentation in MFEM multiapps and timesteppers tests, and set recover=false. Refs #31017

alexanderianblair committed:
Remove platform restriction from MFEM multiapp and timestepper test specifications. Refs #31017

alexanderianblair committed:
Add missing quote mark in MFEM multiapps tests. Refs #31017

alexanderianblair committed:
Move MFEMFullSolveMultApp tests outside MFEMTransientMultiApp test specification. Refs #31017

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3038341/), step Generate coverage on 3149cfd wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9ec918</th>
      <th colspan=4>#31018 3149cf</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.53%</td>
      <td>85.53%</td>
      <td>-0.00%</td>
      <td>94.81%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>115343</td>
      <td>115359</td>
      <td>+16</td>
      <td>73</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19516</td>
      <td>19521</td>
      <td>+5</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31018/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31018/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31018/coverage/xfem)

</details>



This comment will be updated on new commits.

alexanderianblair committed:
Fix comments in MFEM executioner headers. Refs #31017

alexanderianblair committed:
Fix bug affecting transient MFEM problems where all applied kernels are time dependent. Refs #31017

lindsayad reviewed (commented):
I think we should try to avoid `MFEMSteady` and `MFEMTransient` having the multiple inheritance like you pointed out. It has a bit of a code smell and that can tend to runaway down a slippery slope. This has the feel of something that's easier in the short-term but that could be a source of technical debt for which the maintenance cost ends up exceeding the short-term cost savings

alexanderianblair commented:
I don't disagree... OK, let me try and refactor things to turn `MFEMExecutioner` into something like an `MFEMProblemSolve`...

alexanderianblair committed:
Remove multiple inheritance from MFEMTransient and MFEMSteady. Refs #31017

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3053673/), step Clang format on 1f57340 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31018/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31018/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2bf80888ed8064ebb4f92fba1db5c743ac2be9dd`

alexanderianblair committed:
Makes remaining MFEMSteady attributes private and reorders member initialisation. Refs #31017

alexanderianblair committed:
Fix missing _time_step reference in MFEMSteady. Refs #31017

alexanderianblair committed:
Removes now unused output_iteration_number from MFEMSteady. Refs #31017

alexanderianblair committed:
Remove MFEMExecutioner.h from MFEMProblem header. Refs #31017

alexanderianblair committed:
Rename MFEMExecutioner to MFEMProblemSolve to better reflect new class. Refs #31017

alexanderianblair committed:
Remove initProblemOperator from MFEMProblem. Refs #31017

alexanderianblair committed:
Rename ProblemOperatorInterface and EquationSystemInterface to ProblemOperatorBase and MFEMEquationSystemInterface to better reflect their usage. Refs #31017

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3060050/), step Clang format on 0d89f37 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31018/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31018/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b33c61267d17833188a088dff2a8a21a29e957ef`

alexanderianblair committed:
Add default ProblemOperators to Executioners in MFEMProblem::addProblemOperator. Refs #31017

alexanderianblair committed:
Loop over ProblemOperators in MFEM executioners. Refs #31017

alexanderianblair committed:
Update MFEMProblemSolve to inherit from SolveObject. Refs #31017

moosebuild commented:
Job [Clang](https://civet.inl.gov/job/3065882/) on 1254651 : invalidated by @loganharbour

alexanderianblair committed:
Set MFEM convergence flags consistently from MFEMProblemSolve::solve(). Refs #31017

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3066392/), step Clang format on 7915f64 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31018/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31018/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e82beba543d0047b0777b6f322a16cfa69a50ec9`

alexanderianblair committed:
Apply clang-format to MFEMProblemSolve and MFEMSteady. Refs #31017

alexanderianblair committed:
Apply clang-format patch to TimeDomainProblemOperator. Refs #31017

alexanderianblair committed:
Move defualt ProblemOperator construction back to MFEM executioners. Refs #31017

alexanderianblair committed:
Apply clang format patch. Refs #31017

alexanderianblair committed:
Add executioner preSolve() and postSolve() calls into MFEMProblemSolve. Refs #31017

alexanderianblair committed:
Update PassKey type in MFEMProblemSolve after rebase. Refs #31017

lindsayad commented:
> similar to how native MOOSE

Hey this is native MOOSE now too 😄

alexanderianblair committed:
Apply suggestions from code review. Refs #31017

Use shorthand format for linking to markdown documentation pages of MFEMProblemSolve, MFEMTransient, and MFEMSteady.

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

alexanderianblair committed:
Update inherited MFEMTransient::relativeSolutionDifferenceNorm method following rebase

alexanderianblair committed:
Only check for steady state convergence in TransientMultiApp if _detect_steady_state is true. Refs #31017

alexanderianblair commented:
Minor update of `TransientMultiApp` required here following rebase, as `TransientBase::convergedToSteadyState` was being called in `TransientMultiApp::solveStep` even when `_detect_steady_state` is false - appears this was introduced in [#31010](https://github.com/idaholab/moose/pull/31010)

lindsayad commented on framework/include/mfem/interfaces/MFEMProblemOperatorInterface.h:
```diff
@@ -0,0 +1,46 @@
+  virtual ~MFEMProblemOperatorInterface() = default;
+
+  /// Returns a pointer to the operator's equation system.
+  virtual std::vector<std::shared_ptr<Moose::MFEM::ProblemOperatorBase>> & getProblemOperators()
+  {
+    return _problem_operators;
+  }
+
+  /// Add an MFEM problem operator. Takes ownership.
+  virtual void
+  addProblemOperator(std::shared_ptr<Moose::MFEM::ProblemOperatorBase> problem_operator)
+  {
+    _problem_operators.push_back(std::move(problem_operator));
+  }
+
+private:
+  std::vector<std::shared_ptr<Moose::MFEM::ProblemOperatorBase>> _problem_operators;
```
within the namespace we should be able to remove the `Moose::MFEM` qualification?

alexanderianblair replied:
Indeed - removed

lindsayad commented on framework/include/mfem/problem/MFEMProblem.h:
```diff
@@ -71,9 +69,9 @@ class MFEMProblem : public ExternalProblem
+   * Add a ProblemOperator to the executioner used to solve the problem.
    */
-  void initProblemOperator();
+  void addProblemOperator() {};
```
```suggestion
  void addProblemOperator() {}
```

lindsayad commented on framework/include/mfem/problem/MFEMProblem.h:
```diff
@@ -71,9 +69,9 @@ class MFEMProblem : public ExternalProblem
 
   /**
-   * Initialise the required ProblemOperator used in the Executioner to solve the problem.
+   * Add a ProblemOperator to the executioner used to solve the problem.
```
this no longer does what the doxygen says it does

alexanderianblair replied:
Now removed... my hope is that we end up adding this method back in to support custom `ProblemOperators` (that may be useful to specify predefined physics, for example), but realistically this will require a change in the associated `Action` and new `MooseObjects` to support this - outside the scope of this PR.

lindsayad commented on framework/include/mfem/problem_operators/TimeDomainEquationSystemProblemOperator.h:
```diff
@@ -11,27 +11,27 @@
     : TimeDomainProblemOperator(problem),
-      _equation_system(
-          std::dynamic_pointer_cast<Moose::MFEM::TimeDependentEquationSystem>(problem.eqn_system))
+      _equation_system(std::dynamic_pointer_cast<Moose::MFEM::TimeDependentEquationSystem>(
```
```suggestion
      _equation_system(std::dynamic_pointer_cast<TimeDependentEquationSystem>(
```

lindsayad commented on framework/include/mfem/problem_operators/TimeDomainProblemOperator.h:
```diff
@@ -11,22 +11,21 @@
 
   void SetGridFunctions() override;
-
+  void Solve() override {};
```
```suggestion
  void Solve() override {}
```

lindsayad commented on framework/src/mfem/equation_systems/EquationSystem.C:
```diff
@@ -488,22 +488,26 @@ TimeDependentEquationSystem::BuildBilinearForms()
-                                    *(*markers[i]));
-    }
+      for (int i = 0; i < integs->Size(); ++i)
+      {
+        sum->AddIntegrator(*integs[i]);
+      }
```
```suggestion
        sum->AddIntegrator(*integs[i]);
```

lindsayad commented on framework/src/mfem/equation_systems/EquationSystem.C:
```diff
@@ -488,22 +488,26 @@ TimeDependentEquationSystem::BuildBilinearForms()
-    // scaled_sum is owned by td_blf
-    td_blf->AddDomainIntegrator(scaled_sum);
+      for (int i = 0; i < b_integs->Size(); ++i)
+      {
+        td_blf->AddBoundaryIntegrator(new ScaleIntegrator(*b_integs[i], _dt_coef.constant, false),
+                                      *(*markers[i]));
+      }
```
```suggestion
        td_blf->AddBoundaryIntegrator(new ScaleIntegrator(*b_integs[i], _dt_coef.constant, false),
                                      *(*markers[i]));
```

lindsayad commented on framework/src/mfem/executioners/MFEMProblemSolve.C:
```diff
@@ -0,0 +1,98 @@
+  _mfem_problem.updateActiveObjects();
+
+  if (_mfem_problem.shouldSolve())
+  {
+    for (const auto & problem_operator : _problem_operators)
+    {
+      problem_operator->Solve();
+    }
+  }
```
```suggestion
    for (const auto & problem_operator : _problem_operators)
      problem_operator->Solve();
```

lindsayad commented on framework/src/mfem/problem_operators/TimeDomainEquationSystemProblemOperator.C:
```diff
@@ -30,6 +30,24 @@ TimeDomainEquationSystemProblemOperator::Init(mfem::BlockVector & X)
+  _problem_data.ode_solver->Step(_problem_data.f, _problem.time(), _problem.dt());
+  // Synchonise time dependent GridFunctions with updated DoF data.
+  SetTestVariablesFromTrueVectors();
+  // Sync Host/Device
+  _problem_data.f.HostRead();
```
In #31064 we are removing this line

lindsayad commented on test/tests/mfem/multiapps/dt_from_parent.i:
```diff
@@ -0,0 +1,74 @@
+[]
+
+[Variables]
+  [./u]
+  [../]
```
```suggestion
  [u]
  []
```

lindsayad commented on test/tests/mfem/multiapps/dt_from_parent.i:
```diff
@@ -0,0 +1,74 @@
+[]
+
+[Kernels]
+  [./diff]
+    type = MFEMDiffusionKernel
+    variable = u
+  [../]
+  [./td]
+    type = MFEMTimeDerivativeMassKernel
+    variable = u
+  [../]
+[]
+
+[BCs]
+  [./left]
+    type = MFEMScalarDirichletBC
+    variable = u
+    boundary = left
+    coefficient = 0
+  [../]
+  [./right]
+    type = MFEMScalarDirichletBC
+    variable = u
+    boundary = right
+    coefficient = 1
+  [../]
```
```suggestion
  [diff]
    type = MFEMDiffusionKernel
    variable = u
  []
  [td]
    type = MFEMTimeDerivativeMassKernel
    variable = u
  []
[]

[BCs]
  [left]
    type = MFEMScalarDirichletBC
    variable = u
    boundary = left
    coefficient = 0
  []
  [right]
    type = MFEMScalarDirichletBC
    variable = u
    boundary = right
    coefficient = 1
  []
```

lindsayad commented on test/tests/mfem/multiapps/dt_from_parent.i:
```diff
@@ -0,0 +1,74 @@
+[Solver]
+  type = MFEMHypreGMRES
+  preconditioner = boomeramg
+  l_tol = 1e-16
```
does the tol need to be this tight?

alexanderianblair replied:
It does not - relaxed

lindsayad commented on test/tests/mfem/multiapps/dt_from_parent_sub.i:
```diff
@@ -0,0 +1,66 @@
+[]
+
+[Variables]
+  [./u]
```
remove `./` and `../` syntax plz

lindsayad commented on test/tests/mfem/multiapps/tests:
```diff
@@ -0,0 +1,50 @@
+                  OutputData/dt_from_parent/Run0/Cycle000010/proc000000.vtu
+                  OutputData/dt_from_parent_sub/Run0/Run0.pvd
+                  OutputData/dt_from_parent_sub/Run0/Cycle000009/proc000000.vtu'
+      detail = "with a time step governed by the MFEM parent application."
```
```suggestion
      detail = "with a time step governed by the MFEM parent application,"
```

lindsayad commented on test/tests/mfem/multiapps/tests:
```diff
@@ -0,0 +1,50 @@
+                  OutputData/sub_cycling_parent/Run0/Cycle000002/proc000000.vtu
+                  OutputData/sub_cycling_sub/Run0/Run0.pvd
+                  OutputData/sub_cycling_sub/Run0/Cycle000002/proc000000.vtu'
+      detail = "with the sub-app running multiple timesteps for each timestep of the parent app."
```
```suggestion
      detail = "and with the sub-app running multiple timesteps for each timestep of the parent app."
```

lindsayad commented on test/tests/mfem/multiapps/tests:
can we add `devices = 'cpu cuda'` for all of these?

alexanderianblair replied:
Yes - added

lindsayad commented on test/tests/mfem/timesteppers/tests:
```diff
@@ -0,0 +1,27 @@
+      type = CSVDiff
+      input = 'mfem_multiple_timesteppers.i'
+      csvdiff = 'mfem_multiple_timesteppers.csv'
+      detail = 'using the minimum time step size of input time steppers.'
```
```suggestion
      detail = 'using the minimum time step size of input time steppers,'
```

lindsayad commented on test/tests/mfem/timesteppers/tests:
```diff
@@ -0,0 +1,27 @@
+      type = CSVDiff
+      input = 'mfem_multiple_timesequences.i'
+      csvdiff = 'mfem_multiple_timesequences.csv'
+      detail = 'with specific times to hit and still hit all of them.'
```
```suggestion
      detail = 'or with specific times to hit.'
```

lindsayad commented on test/tests/mfem/timesteppers/tests:
can we set `devices = 'cpu cuda'` for these?

alexanderianblair committed:
Apply formatting suggestions from code review

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3086637/), step Clang format on c499f54 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31018/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31018/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0aecda05dbe7d2aeeac414d2adbdad7949df51e9`

alexanderianblair committed:
Relax tolerances of MFEM timestepper and multiapp tests and add cuda to test devices. Refs #31017

alexanderianblair committed:
Rename MFEMProblemOperatorInterface and MFEMEquationSystemInterface to ProblemOperatorInterface and EquationSystemInterface respectively. Refs #31017

alexanderianblair committed:
Remove Device and unnecessary includes from MFEMProblemSolve. Refs #31017

alexanderianblair committed:
Remove hostread from TimeDomainEquaionSystemProblemOperator::Solve. Refs #31017

alexanderianblair commented on framework/include/mfem/executioners/MFEMProblemSolve.h:
```diff
@@ -0,0 +1,43 @@
+protected:
+  MFEMProblem & _mfem_problem;
+  std::vector<std::shared_ptr<Moose::MFEM::ProblemOperatorBase>> & _problem_operators;
+  mfem::Device _device;
```
Indeed - removed

alexanderianblair commented on framework/include/mfem/interfaces/MFEMEquationSystemInterface.h:
```diff
@@ -19,11 +19,11 @@ namespace Moose::MFEM
  * - TimeDomainEquationSystemProblemOperator
  */
-class EquationSystemInterface
+class MFEMEquationSystemInterface
```
I've now removed the MFEM prefix from `EquationSystemInterface` and `ProblemOperatorInterface`, as I agree that doubling up doesn't make sense.

I assume we would probably want all classes for MFEM problem setup to belong to the `Moose::MFEM::` namespace if they aren't a registered MooseObject for creation by the user? As currently the other interfaces (`MFEMBlockRestrictable` and `MFEMBoundaryRestrictable` use the prefix instead

lindsayad replied:
> I assume we would probably want all classes for MFEM problem setup to belong to the `Moose::MFEM::` namespace if they aren't a registered MooseObject for creation by the user?

Yes I agree with this

alexanderianblair commented on framework/include/mfem/interfaces/MFEMProblemOperatorInterface.h:
```diff
@@ -0,0 +1,46 @@
+ * - MFEMTransient
+ * - MFEMSteady
+ */
+class MFEMProblemOperatorInterface
```
As above - changed to ProblemOperatorInterface

alexanderianblair committed:
Remove unused addProblemOperator from MFEMProblem. Refs #31017

alexanderianblair committed:
Add comments explaining max_parallel limits in new test specifications

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

alexanderianblair committed:
Remove max_threads constraint from new MFEM timestepper and multiapp tests. Refs #31017

alexanderianblair committed:
Add missing Executioner header into MFEMSteady.h. Refs #31017

alexanderianblair commented on test/tests/mfem/timesteppers/tests:
```diff
@@ -0,0 +1,29 @@
+      devices = 'cpu cuda'
+      max_parallel = 1
+      recover = false
+      max_threads = 1
```
Did some digging - looks like `max_threads` is safe to remove in these tests (now removed). They do appear to be required for the curl curl examples or the `gravity.i` test, though - notably those use hypre preconditioners with options that other tests don't.
Recovery isn't supported yet - although this PR does help bring it closer!

alexanderianblair committed:
Add device parameter explicitly to new MFEM test input files. Refs #31017

alexanderianblair committed:
Add MFEM FESpace specification to timestepper and multiapp tests. Refs #31017

nmnobre reviewed (commented):
Excellent work, I like these changes a lot :)

alexanderianblair committed:
Apply documentation fixes from code review

Co-authored-by: Nuno Nobre <nuno.nobre@stfc.ac.uk>

alexanderianblair commented on framework/src/mfem/executioners/MFEMSteady.C:
```diff
@@ -11,37 +11,39 @@
-    _output_iteration_number(0)
+    _last_solve_converged(false)
 {
   _time = _system_time;
```
Hesitant to remove this line to maintain consistency with similar usage in `Steady` and `SteadyBase` here

nmnobre replied:
Ah, I see... On the flip side, we'd gain some nice consistency with `MFEMTransient` ^^

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3095618/), step Clang format on fa61053 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31018/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31018/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ee92e3336165cdfa3c620db4cd7aa3d656fc2c9c`

alexanderianblair committed:
Apply clang-format to MFEMProblemSolve.C. Refs #31017

sean-baccas committed:
Change solver to HyprePCG to avoid cuda issue

alexanderianblair committed:
Update MFEM multiapp and timesteppers test specifications to use compute_devices following rebase. Refs #31017

alexanderianblair committed:
Disable restep testing for MFEMTransient executioner tests. Refs #31017

lindsayad commented:
In the future let's try to squash the clang formatting commits

alexanderianblair commented:
> In the future let's try to squash the clang formatting commits

Will do
